File: MicroBenchmarks/Py/main_lapl_2d.py
import numpy as np
import time

def Init(X,L):
    size=X.shape[0]
    h=L/size
    for i in range(0,size):
        for j in range(0,size):
            if i>size//8 and i<size//2+size//8:
                X[i]=1.-2*(i-size//8)*h/L;
            else:
                X[i]=0.0


def lapl2d_1(In,Out,niter):
    siz=In.shape[0]
    h2= (1./size)**2

    for it in range(0,niter):
        Out[1:siz-1,1:siz-1]= h2*(
            In[0:siz-2,1:siz-1 ] + In[1:siz-1,0:siz-2]-
            4.0*In[1:siz-1,1:siz-1]+
            In[2:siz,1:siz-1]+In[1:siz-1,2:siz])
        In,Out=Out,In

def lapl2d_2(In,Out,niter):
    size=In.shape[0]
    h2= (1./size)**2
    for it in range(0,niter):
        for i in range(1,size-1):
            for j in range(1,size-1):
                Out[i,j]= h2*(
                    In[i-1,j] + In[i,j-1]-
                    4.0*In[i,j]+
                    In[i+1,j]+In[i,j+1])
        In,Out=Out,In  
      
def test(p,In,Out,nit):
   
    niter=nit
    Init(In,1.)
    Init(Out,1.)
 
    p(In,Out,niter)
    T=0.
    while True:
        Init(In,1.)
        Init(Out,1.)
        t1 = time.time()
        p(In,Out,niter)
        treal=time.time() -t1
        t = treal/niter
        if treal>0.0001 and abs(t-T)/t<0.025:
            break
        else:
            T=t
            niter*=2

    return T,niter

size=16
sizemax=513
niter=10
parsef= lambda  f: str(f).split(" ")[1] #parse function name
while size<sizemax:
    print("size: ",size)
    In= np.empty((size,size))
    Out= np.empty((size,size))
    tbest=10.**20
    best=0
    t=0.0
    for p in  [lapl2d_1,lapl2d_2]:
        t,it=test(p,In,Out,niter)
        if t<tbest:
            tbest=t
            best=p
        print(parsef(p)," : t= ",t," seconds ")
    nflops=6*(size-2)**2
    flops=nflops/tbest
    print("\nbest: ",parsef(best))
    print("nb. flops (best): ",nflops, ", Gflops/s: ",flops/(10**9))
    print("-------")
    size*=2
    print(" ")

File: MicroBenchmarks/Ju/main_lapl_2d.jl

# initalize an array
function Init!(X::Array{Float64,2},L)
    siz=size(X)[1]
    h=L/siz
    for j=0:siz-1
        for i=0:siz-1
            if i>floor(siz//8) && i<floor(siz//2)+floor(siz//8)
                X[i+1,j+1]=1.-2*(i-floor(siz/8))*h/L;
            else
                X[i+1,j+1]=0.0
            end
        end
    end
end



# proc[1-5]! functions make the same computation, with different programing
# styles.

function proc1!(In::Array{Float64,2},Out::Array{Float64,2},niter::Int64)
    const siz=size(In)[1]
    h2::Float64= (1./siz)^2
    for it=1:niter
        Out[2:siz-1,2:siz-1]=
        h2*(In[1:siz-2,2:siz-1 ] + In[2:siz-1,1:siz-2]-
            4.0*In[2:siz-1,2:siz-1]+
            In[3:siz,2:siz-1]+In[2:siz-1,3:siz])
        In,Out=Out,In
    end 
end
function proc2!(In::Array{Float64,2},Out::Array{Float64,2},niter::Int64)
    const siz=size(In)[1]
    h2::Float64= (1./siz)^2
    for it=1:niter
       @. Out[2:siz-1,2:siz-1]=
        h2*(In[1:siz-2,2:siz-1 ] + In[2:siz-1,1:siz-2]-
            4.0*In[2:siz-1,2:siz-1]+
            In[3:siz,2:siz-1]+In[2:siz-1,3:siz])
        In,Out=Out,In
    end 
end
function proc3!(In::Array{Float64,2},Out::Array{Float64,2},niter::Int64)
    const siz=size(In)[1]
    h2::Float64= (1./siz)^2
    for it=1:niter
        for j=2:siz-1
            for i=2:siz-1
                Out[i,j]=
                h2*(In[i-1,j] + In[i,j-1]-
                    4.0*In[i,j]+
                    In[i+1,j]+In[i,j+1])
            end
        end
        In,Out=Out,In
    end 
end
function proc4!(In::Array{Float64,2},Out::Array{Float64,2},niter::Int64)
    const siz=size(In)[1]
    h2::Float64= (1./siz)^2
    for it=1:niter
        for j=2:siz-1
            @simd for i=2:siz-1
                Out[i,j]=
                h2*(In[i-1,j] + In[i,j-1]-
                    4.0*In[i,j]+
                    In[i+1,j]+In[i,j+1])
            end
        end
        In,Out=Out,In
    end 
end
function proc5!(In::Array{Float64,2},Out::Array{Float64,2},niter::Int64)
    const siz=size(In)[1]
    h2::Float64= (1./siz)^2
    for it=1:niter
        for i=2:siz-1
            @simd for j=2:siz-1
                Out[i,j]=
                h2*(In[i-1,j] + In[i,j-1]-
                    4.0*In[i,j]+
                    In[i+1,j]+In[i,j+1])
            end
        end
        In,Out=Out,In
    end 
end
# measure as precisely as possible the execution time for a given
# p= proc[1-5]!
function test(p,In::Array{Float64,2},Out::Array{Float64,2},nit::Int)
   
    niter=nit
    Init!(In,1.)
    Init!(Out,1.)
 
    p(In,Out,niter)
    T=0.
    while true
        Init!(In,1.)
        Init!(Out,1.)
        t1 = time_ns()
        p(In,Out,niter)
        t = (time_ns() -t1)/niter
        if abs(t-T)/t<0.025
            break
    else
            T=t
            if niter> 150000
                error("too many iterations",niter)
            else
                niter*=2
            end
    end
    
end
    T,niter
end
DD=Dict("proc1!"=>"Vectorisation with braces             ",
        "proc2!"=>"Vectorisation with @. and braces      ",
        "proc3!"=>"Unrolled loop (good order)            ",
        "proc4!"=>"Unrolled loop with @simd (good order) ",
        "proc5!"=>"Unrolled loop with @simd (bad order)  "
        )
# computation starts here:
siz=8
sizemax=1025
const niter=2
while siz<sizemax
    println("siz: ",siz)
    In=Array{Float64,2}(siz,siz)
    Out=Array{Float64,2}(siz,siz)
    tbest=10.^29
    best=0
    t=0.0
    for p= [proc1!,proc2!,proc3!,proc4!,proc5!]
        t,it=test(p,In,Out,niter)
        if t<tbest
            tbest=t
            best=p
        end
        t*=10.0^(-9)
        println(DD[string(p)]," : t= ",t," seconds ")
    end

    nflops=6*(siz-2)^2
    flops=nflops/tbest
    println("\nbest: ",DD[string(best)])
    println("nb. flops (best): ",nflops, ", Gflops/s: ",flops)
    println("-------")
    
    siz*=2
    println()
end
